### ISSUE TITLE
Speeding up BoundaryAttack and HopSkipJump by increasing the batch_size

### ISSUE THREAD

**Issue by amodas** [Trusted-AI/adversarial-robustness-toolbox#970]
### Issue

In their current implementation, both [`BoundaryAttack`](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/evasion/boundary.py) and [`HopSkipJump`](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/evasion/hop_skip_jump.py) have a hardcoded `batch_size=1` in their code ([L107](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/3a957076d0df87203e1056b442a59d4ff56a8810/art/attacks/evasion/boundary.py#L107) and [L96](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/3a957076d0df87203e1056b442a59d4ff56a8810/art/attacks/evasion/hop_skip_jump.py#L96) respectively). In fact, `batch_size` is used to inform the classifier how many samples to fit/forward simultaneously, and then return the predictions for all the samples within the batch.

However, the hardcoded value `batch_size=1` is inefficient: whenever the attacks want to get the predictions of `N` samples, the classifier is feeding them one-by-one instead of feeding all `N` samples simultaneously. As a result, the whole forward process is quite slow, and the main benefit of parallelism in a GPU is technically lost!

I studied the code of both implementations to make sure that there is no **specific reason** (cases such as vectorization of the returned predictions etc.) for explicitly using `batch_size=1`. By experimenting **only** with PyTorch, it turned out that both algorithms **should work as expected for any `batch_size >= 1`**. I also believe that changing the backend (i.e., Tensorflow, Keras) will not change the behaviour (at least I do not see any reason why).

To demonstrate the benefit of setting `batch_size > 1`, I created a simple experiment where I deterministically evaluate both attacks (a) in their current state, and (b) with an increased `batch_size`. The code and results can be found in the Notebook embedded below (see Appendix).

### Results

As you can see from the Notebook below, the running time after increasing the batch size is significantly improved:

1. BoundaryAttack
   - 475 sec (before)
   - 243 sec (after)

2. HopSkipJump
   - 392 sec (before)
   - 228 sec (after)

### Suggested Fix
I will open a pull request in order to integrate this into ART, since I believe that it is safe to modify the code such that:
- The attack has `batch_size` as an argument during `__init__`
- The attack has a default value larger than 1 (i.e., `batch_size: int = 64`)

### Important! 
I would recommend to gradually check if the same situation occurs in other attacks (white- or black-box ones), since such a small change is enough to boost the execution performance of the attack.

# Appendix – Notebook

**Note**: the experiments use PyTorch v1.6.0+cu92 and Torchvision v0.7.0+cu92 since they are conducted on a Tesla K40 GPU. The overall behaviour is not expected to change if tested with higher version of PyTorch and Torchvision.

```python
global_seed = 999

import sys
import sys
!{sys.executable} -m pip install git+https://github.com/nottombrown/imagenet_stubs
sys.path.append('./adversarial-robustness-toolbox')

import imagenet_stubs
import numpy as np
import torch
import torch.nn as nn
import torchvision.models as models
import torchvision.transforms as transforms
import random
import time

import os
os.environ['PYTHONHASHSEED'] = str(global_seed)

from matplotlib import pyplot as plt
from PIL import Image

from art.estimators.classification import PyTorchClassifier
from art.attacks.evasion import BoundaryAttack, HopSkipJump
```

    Collecting git+https://github.com/nottombrown/imagenet_stubs
      Cloning https://github.com/nottombrown/imagenet_stubs to /tmp/pip-req-build-fr1ufr_e
      Running command git clone -q https://github.com/nottombrown/imagenet_stubs /tmp/pip-req-build-fr1ufr_e

### Set determinism for reproducibility

```python
def set_global_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.manual_seed(seed)
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True
```

### Load Torchvision model

```python
model = models.resnet18(pretrained=True).eval()

mean = np.asarray((0.485, 0.456, 0.406), dtype=np.float32).reshape((1, 3, 1, 1))
std = np.asarray((0.229, 0.224, 0.225), dtype=np.float32).reshape((1, 3, 1, 1))

classifier = PyTorchClassifier(
    model=model,
    clip_values=(0.0, 1.0),
    input_shape=(3, 224, 224),
    nb_classes=1000,
    use_amp=False,
    preprocessing=(mean, std),
    channels_first=True,
    loss=nn.CrossEntropyLoss()
)
```

### Define image transformation

```python
def tf(x):
    x = np.expand_dims(x, axis=0)
    return np.transpose(x, (0, 3, 1, 2))
```

### Load image

```python
original_image_name = 'standard_poodle.jpg'
for image_path in imagenet_stubs.get_image_paths():
    if image_path.endswith(original_image_name):
        original_image = Image.open(image_path)
        original_image = original_image.resize((224, 224))
        original_image = np.array(original_image) / 255.0
        original_image = original_image.astype(np.float32)

original_label = np.argmax(classifier.predict(tf(original_image)), axis=1)[0]
print("\nOriginal label is: %d\n" % original_label)

plt.imshow(original_image)
plt.show()
```

    
    Original label is: 267
   
    
![output_8_1](https://user-images.githubusercontent.com/25585560/111333254-d00f7600-8672-11eb-92b0-1878a3421058.png)

    

### BoundaryAttack parameters

```python
max_iterations = 1000
delta = 0.01
epsilon = 0.01
targeted = False
```

### BoundaryAttack (`batch_size=1`)

```python
set_global_seed(global_seed)

batch_size = 1  # <--- Current implementation (hardcoded) batch size
attack = BoundaryAttack(estimator=classifier,
                        targeted=targeted,
                        max_iter=max_iterations,
                        delta=delta,
                        epsilon=epsilon,
                        verbose=False,
                        batch_size=batch_size,
                        seed=global_seed)

t0 = time.time()
x_adv = attack.generate(x=tf(original_image), x_adv_init=None)
print('Elapsed time (sec):', time.time() - t0)

L2_err = np.linalg.norm(x_adv[0] - tf(original_image)[0])
pred_label = np.argmax(classifier.predict(x_adv)[0])
print("L2 error: %.2f \t | \t Adv. label: %d" % (L2_err, pred_label))
```

    Elapsed time (sec): 475.44113087654114
    L2 error: 31.02 	 | 	 Adv. label: 107

### BoundaryAttack (`batch_size=128`)

```python
set_global_seed(global_seed)

max_iterations = 1000
delta = 0.01
epsilon = 0.01
targeted = False

batch_size = 128  # <--- Increase batch size

attack = BoundaryAttack(estimator=classifier,
                        targeted=targeted,
                        max_iter=max_iterations,
                        delta=delta,
                        epsilon=epsilon,
                        verbose=False,
                        batch_size=batch_size,
                        seed=global_seed)

t0 = time.time()
x_adv = attack.generate(x=tf(original_image), x_adv_init=None)
print('Elapsed time (sec):', time.time() - t0)

L2_err = np.linalg.norm(x_adv[0] - tf(original_image)[0])
pred_label = np.argmax(classifier.predict(x_adv)[0])
print("L2 error: %.2f \t | \t Adv. label: %d" % (L2_err, pred_label))
```

    Elapsed time (sec): 243.049170255661
    L2 error: 31.02 	 | 	 Adv. label: 107

### HopSkipJump Attack parameters

```python
max_iterations = 200
max_eval = 1000
init_eval = 10
targeted = False
```

### HopSkipJump Attack (`batch_size=1`)

```python
set_global_seed(global_seed)

batch_size = 1  # <--- Current implementation (hardcoded) batch size
attack = HopSkipJump(classifier=classifier,
                     targeted=targeted,
                     max_iter=max_iterations,
                     max_eval=max_eval,
                     init_eval=init_eval,
                     verbose=False,
                     batch_size=batch_size,
                     seed=global_seed)

t0 = time.time()
x_adv = attack.generate(x=tf(original_image), x_adv_init=None)
print('Elapsed time (sec):', time.time() - t0)

L2_err = np.linalg.norm(x_adv[0] - tf(original_image)[0])
pred_label = np.argmax(classifier.predict(x_adv)[0])
print("L2 error: %.2f \t | \t Adv. label: %d" % (L2_err, pred_label))
```

    Elapsed time (sec): 392.41476559638977
    L2 error: 0.96 	 | 	 Adv. label: 183

### HopSkipJump Attack (`batch_size=128`)

```python
set_global_seed(global_seed)

batch_size = 128  # <--- Increase batch size
attack = HopSkipJump(classifier=classifier,
                     targeted=targeted,
                     max_iter=max_iterations,
                     max_eval=max_eval,
                     init_eval=init_eval,
                     verbose=False,
                     batch_size=batch_size,
                     seed=global_seed)

t0 = time.time()
x_adv = attack.generate(x=tf(original_image), x_adv_init=None)
print('Elapsed time (sec):', time.time() - t0)

L2_err = np.linalg.norm(x_adv[0] - tf(original_image)[0])
pred_label = np.argmax(classifier.predict(x_adv)[0])
print("L2 error: %.2f \t | \t Adv. label: %d" % (L2_err, pred_label))
```

    Elapsed time (sec): 228.36624002456665
    L2 error: 0.97 	 | 	 Adv. label: 183
